File: Problems/Dynamic-programming/Longest-Common-Subsequence/Longest-Common-Subsequence.cs
using System;

public class Solution
{
    public int LongestCommonSubsequence(string text1, string text2)
    {
        int m = text1.Length;
        int n = text2.Length;

        if (m < n)
        {
            (m, n) = (n, m);
            (text1, text2) = (text2, text1);
        }

        int[] dp = new int[n + 1];

        for (int i = 0; i < m; i++)
        {
            int prev = 0;

            for (int j = 0; j < n; j++)
            {

                int cur = dp[j + 1];
                if (text1[i] == text2[j])
                {
                    dp[j + 1] = prev + 1;
                }
                else
                {
                    dp[j + 1] = Math.Max(dp[j], dp[j + 1]);
                }
                prev = cur;
            }
        }

        return dp[n];
    }
}

class Program
{
    static void Main()
    {
        string text1 = "abcde";
        string text2 = "ace";

        Console.WriteLine("Input: text1 = " + text1 + ", text2 = " + text2);

        Solution sol = new Solution();
        int result = sol.LongestCommonSubsequence(text1, text2);

        Console.WriteLine("Output: " + result);
    }
}

File: Problems/Dynamic-programming/Longest-Common-Subsequence/Longest-Common-Subsequence.cpp
#include <iostream>
#include <vector>
#include <string>

// Using tabulation - Time: O(m*n)

class Solution {
public:
    int longestCommonSubsequence(std::string text1, std::string text2)
    {
        int m = text1.size(), n = text2.size();

        if (m < n)
            std::swap(m, n), swap(text1, text2);

        std::vector<int> dp(n + 1);

        for (int i = 0; i < m; i++)
        {
            int prev = 0;

            for (int j = 0; j < n; j++)
            {
                int cur = dp[j + 1];
                if (text1[i] == text2[j])
                {
                    dp[j + 1] = prev + 1;
                }
                else
                {
                    dp[j + 1] = std::max(dp[j], dp[j + 1]);
                }
                prev = cur;
            }
        }

        return dp[n];
    }
};

int main()
{
    std::string text1 = "abcde";
    std::string text2 = "ace";

    std::cout << "Input: text1 = " << text1 << ", text2 = " << text2 << std::endl;

    Solution sol;
    int result = sol.longestCommonSubsequence(text1, text2);

    std::cout << "Output: " << result << std::endl;
    
    return 0;
}
